### ISSUE TITLE
Restoring workspace containing stanfit object destabilizes R

### ISSUE THREAD

**Issue by ksvanhorn** [stan-dev/rstan#273]
#### Summary:

Restoring a workspace containing a stanfit object destabilizes R when a namespace that was available when the object was created is not available when the workspace is restored.
#### Description:

I'm working in a package in RStudio, call it Foo, and in my environment I have an object w2b that contains several stanfit objects within it (84 in total). I quit my R session and save my workspace. Later, I reopen the package, which restores the workspace, and see this message:

"Warning: namespace 'Foo" is not available and has been replaced by .GlobalEnv when processing object 'w2b'."

And now I find that various functions such as chol(), eigen(), determinant() no longer work. determinant(diag(2)) either crashes the system or returns NULL.

If I remove the object w2b from the environment, quit my R session (saving the workspace), and restart R, reopening the package Foo again, everything is fine.
#### RStan Version:

2.9.0
#### R Version:

3.2.2
#### Operating System:

Windows 7

**Comment by ksvanhorn** [comment]
I've been experimenting with this, and I am baffled by what I'm seeing. I'm not sure this is an rstan-specific bug; it may be an R bug.

Some more specifics:
- The w2b object mentioned above is a list of two objects named lt and ltwp.
- Each of the list elements is itself a list-based matrix of 42 elements.
- Each element of the matrices is a named list containing (among other things) a stanfit object named fit, and a closure named forecast.

I tried various experiments in which I tossed away most of w2b and kept only some subset of the closures and stanfit objects, then carried out the following sequence: shut down R (saving the workspace), then restart it (restoring the workspace), then execute "determinant(diag(2))". This is what I found:
- If I keep just the stanfit objects from either w2b$lt or w2b$ltwp, everything is fine.
- If I keep the stanfit objects from _both_ w2b$lt and w2b$ltwp, R crashes at the end of the test sequence.
- If I keep just the closure objects from w2b$lt or w2b$ltwp, everything is fine.
- If I keep the closure objects from _both_ w2b$lt and w2b$ltwp, R crashes at the end of the test sequence.

**Comment by sakrejda** [comment]
Intersting, I haven't explored the rstan internals very well but from my experience with Rcpp I know saving lists of external pointers and expecting them to still function after restoring them from saved data is touchy.  Even a package without external pointers that saves functions defined in lists can trigger failures (warnings or errors, not outright crashes, but external pointers will convert lots of errors to segfaults, especially if the package is not loaded when the object is).  If this is not something Ben is already familiar with it would help to have a reproducible example to investigate.

**Comment by bgoodri** [comment]
All kinds of weird things happen when S4 objects are loaded or autoloaded into R's workspace. Was this object created with the same version of rstan that is installed currently?

**Comment by ksvanhorn** [comment]
Yes, the version of R and RStan has remained unchanged between saving and restoring the object.
